### ISSUE TITLE
Instead of rising steadily, the reward fluctuates wildly

### ISSUE THREAD

**Issue by Hasuer** [huggingface/open-r1#403]
I used format reward and acc reward. Using the qwen2.5-7b-base model, I noticed that for a given system prompt, the format reward steadily increased, but after a certain number of steps it started to decrease and then started to fluctuate. At the same time, the reward also starts to fluctuate. There's a lot of repetition happening at the same time.

![Image](https://github.com/user-attachments/assets/af91c963-fdfc-45f8-a0f2-cafd780ebd1d)

![Image](https://github.com/user-attachments/assets/cb854ae9-0ccb-44dc-a854-3b9ce2bca363)

The orange line in the picture is the result under $\beta$=0.001 (Coefficient of KL), and the purple line is the result under $\beta$=0. It can be seen that no-zero $\beta$ still cannot avoid making the reward drop to 0, although it can delay the time when the decrease occurs. I also tried different coefficients for KL, but they all gave similar results instead of steadily increasing.

I have a few questions:
1. What might be the reason why rewards fluctuate so much?
2. How can I fix it?

Any comments would really help me!!

**Comment by Hasuer** [comment]
@qgallouedec Perhaps you have some experience to share? ❤

**Comment by qgallouedec** [comment]
How is your learning rate? Maybe reduce it a bit to stabilize.

**Comment by Hasuer** [comment]
> How is your learning rate? Maybe reduce it a bit to stabilize.

Here is the config:
```
# GRPO trainer config
bf16: true
use_vllm: true
vllm_device: auto
vllm_gpu_memory_utilization: 0.7
do_eval: false
# eval_strategy: steps
# eval_steps: 1
gradient_accumulation_steps: 1
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false
hub_model_id: Qwen2.5-7B-Open-R1-GRPO
hub_strategy: every_save
learning_rate: 2.0e-05
log_completions: true
log_level: info
logging_first_step: true
logging_steps: 1
logging_strategy: steps
lr_scheduler_type: cosine
max_prompt_length: 1024
max_completion_length: 2048
max_steps: -1
num_generations: 7
num_train_epochs: 1
output_dir: data/Qwen2.5-7B-Open-R1-GRPO
overwrite_output_dir: true
per_device_eval_batch_size: 8
per_device_train_batch_size: 8
push_to_hub: true
report_to:
- wandb
reward_funcs:
- custom_acc
# - accuracy
- format
reward_weights:
- 1.0
- 1.0
save_strategy: "epoch"
save_total_limit: 1
seed: 42
warmup_ratio: 0.1
beta: 0.001
```
The `custom_acc`reward and training data are adopted from [Open-Reasoner-Zero](https://github.com/Open-Reasoner-Zero/Open-Reasoner-Zero). But I don't think it the reason of unstabilized training.

Besides, I want to inrease `max_prompt_length` from 512 to 1024, and `max_completion_length` from 1024 to 2048 under `num_generations =16 and per_device_train_batch_size = 16` but causing OOM on worker devices, I guess the reason is [this](https://github.com/huggingface/open-r1/issues/400#issuecomment-2676993480). Am I guessing right?

**Comment by lzy37ld** [comment]
mark

**Comment by Mrw33554432** [comment]
I observed a similar behavior when training Qwen-2.5-14B.

<img width="1705" alt="Image" src="https://github.com/user-attachments/assets/2af06541-2d42-4a2f-86bc-39a7e28bfcfc" />

**Comment by Hasuer** [comment]
> I observed a similar behavior when training Qwen-2.5-14B. The sudden drop at start seem quite the same.
I now try the lr=1e-6, and wait to see the results

**Comment by GithubX-F** [comment]
> > I observed a similar behavior when training Qwen-2.5-14B. The sudden drop at start seem quite the same.
> 
> I now try the lr=1e-6, and wait to see the results
> 
> How is your learning rate and beta for KL? [@Mrw33554432](https://github.com/Mrw33554432)

Hello, I have the same question. Have you solved it?

**Comment by reneliury** [comment]
> I observed a similar behavior when training Qwen-2.5-14B. The sudden drop at start seem quite the same.
> 
> <img alt="Image" width="1705" src="https://private-user-images.githubusercontent.com/102270023/416079814-2af06541-2d42-4a2f-86bc-39a7e28bfcfc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIwMDEyNDgsIm5iZiI6MTc0MjAwMDk0OCwicGF0aCI6Ii8xMDIyNzAwMjMvNDE2MDc5ODE0LTJhZjA2NTQxLTJkNDItNGEyZi04NmJjLTM5YTdlMjhiZmNmYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxNVQwMTA5MDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hYmMyNzYzZWVmNjEyOGRiMDM4OWEyY2ZkYTY1NzZlYTc3ZWIwZTAzYTNhY2YwNjNhMTY4ZDM3MjBjNGRmZWRhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.iRcxFasUW_xPc0hfL8BfMrRt5w9KMV3ddXlRZU4kJeg">

@Mrw33554432  Can you train GRPO on a 14B model using OpenR1? Are you using multi-machine training? I have 4 machines with 8 A100 (40G) GPUs each, and I'm not sure if it's feasible to train a 14B model.

**Comment by ngohoangnam28101995** [comment]
![Image](https://github.com/user-attachments/assets/57073cd2-c788-4ada-bde5-536b8e5fce51) I have the same issue when training. This is my training parameter. I have tried many learning rate (9e-6 to 5e-5)  but reward is dropped at around epoch 100-120.
training_args = GRPOConfig(
    use_vllm=True,
    vllm_gpu_memory_utilization=0.7,
    learning_rate=9e-6,
    adam_beta1=0.9,
    adam_beta2=0.999,
    weight_decay=0.01,
    warmup_ratio=0.1,
    lr_scheduler_type="cosine_with_min_lr",
    optim="adamw_8bit",
    lr_scheduler_kwargs={'min_lr_rate': 0.1},
    logging_steps=1,
    bf16=is_bfloat16_supported(),
    fp16=not is_bfloat16_supported(),
    per_device_train_batch_size=1,
    gradient_accumulation_steps=1,
    num_generations=8,
    max_prompt_length=512,
    max_completion_length=2048,
    num_train_epochs=1,
    max_steps=1000,
    save_steps=1000,
    save_total_limit=1,
    report_to="wandb",
    output_dir="outputs",
    overwrite_output_dir=True,
    temperature=0.7,  # Thêm tham số temperature
    max_grad_norm = 10,
)

**Comment by tdano1** [comment]
> ![Image](https://github.com/user-attachments/assets/57073cd2-c788-4ada-bde5-536b8e5fce51) I have the same issue when training. This is my training parameter. I have tried many learning rate (9e-6 to 5e-5) but reward is dropped at around epoch 100-120. training_args = GRPOConfig( use_vllm=True, vllm_gpu_memory_utilization=0.7, learning_rate=9e-6, adam_beta1=0.9, adam_beta2=0.999, weight_decay=0.01, warmup_ratio=0.1, lr_scheduler_type="cosine_with_min_lr", optim="adamw_8bit", lr_scheduler_kwargs={'min_lr_rate': 0.1}, logging_steps=1, bf16=is_bfloat16_supported(), fp16=not is_bfloat16_supported(), per_device_train_batch_size=1, gradient_accumulation_steps=1, num_generations=8, max_prompt_length=512, max_completion_length=2048, num_train_epochs=1, max_steps=1000, save_steps=1000, save_total_limit=1, report_to="wandb", output_dir="outputs", overwrite_output_dir=True, temperature=0.7, # Thêm tham số temperature max_grad_norm = 10, )

Have you solved this? What dataset are you using?
